src/AssetRegistry/UniswapV2PricingModule.sol

```diff
@@ -0,0 +1,371 @@
+     *      by calculating what the reserves of the pool would be if a profit-maximizing trade is done.
+     *      As such flash-loan attacks are mitigated, where an attacker swaps a large amount of the higher priced token,
+     *      to bring the pool out of equilibrium, resulting in liquidity postitions with a higher share of the most valuable token.
+     * @dev Modification of https://github.com/Uniswap/v2-periphery/blob/master/contracts/libraries/UniswapV2LiquidityMathLibrary.sol#L23
+     */
+    function _getTrustedTokenAmounts(
+        address pair,
+        uint256 trustedPriceToken0,
+        uint256 trustedPriceToken1,
+        uint256 liquidityAmount
+    ) internal view returns (uint256 token0Amount, uint256 token1Amount) {
+        uint256 kLast = feeOn ? IUniswapV2Pair(pair).kLast() : 0;
+        uint256 totalSupply = IUniswapV2Pair(pair).totalSupply();
+
+        // this also checks that totalSupply > 0
+        require(totalSupply >= liquidityAmount && liquidityAmount > 0, "UV2_GV: LIQUIDITY_AMOUNT");
+
+        (uint256 reserve0, uint256 reserve1) = _getTrustedReserves(pair, trustedPriceToken0, trustedPriceToken1);
+
+        return _computeTokenAmounts(reserve0, reserve1, totalSupply, liquidityAmount, kLast);
+    }
+
+    /**
+     * @notice Gets the reserves after an arbitrage moves the price to the profit-maximizing ratio given externally observed trusted price
+     * @param pair Address of the Uniswap V2 Liquidity pool
+     * @param trustedPriceToken0 Trusted price of an amount of Token0 in a given BaseCurrency
+     * @param trustedPriceToken1 Trusted price of an amount of Token1 in a given BaseCurrency
+     * @return reserve0 The reserves of token0 in the liquidity pool after arbitrage
+     * @return reserve1 The reserves of token1 in the liquidity pool after arbitrage
+     * @dev Both trusted prices must be for the same BaseCurrency, and for an equal amount of tokens
+     *      e.g. if trustedPriceToken0 is the USD price for 10**18 tokens of token0,
+     *      than trustedPriceToken2 must be the USD price for 10**18 tokens of token1.
+     *      The amount of tokens should be big enough to guarantee enough precision for tokens with small unit-prices
+     * @dev Modification of https://github.com/Uniswap/v2-periphery/blob/master/contracts/libraries/UniswapV2LiquidityMathLibrary.sol#L23
+     */
+    function _getTrustedReserves(address pair, uint256 trustedPriceToken0, uint256 trustedPriceToken1)
+        internal
+        view
+        returns (uint256 reserve0, uint256 reserve1)
+    {
+        // The untrusted reserves from the pair, these can be manipulated!!!
+        (reserve0, reserve1,) = IUniswapV2Pair(pair).getReserves();
+
+        require(reserve0 > 0 && reserve1 > 0, "UV2_GV: ZERO_PAIR_RESERVES");
+
+        // Compute how much to swap to balance the pool with externally observed trusted prices
```

Comment:
Is it possible to early terminate if the pools are balanced

Reply:
it's going to be hard to be exactly balanced, the oracle prices of the underlying assets will never be exactly the ratio of the reserves